bcgov/GDX-Analytics-microservice
# Gdxdsd 7197 derived assets to redshift split

### This PR does the following:

1. Moves and/or copies the python scripts asset_data_to_redshift.py and build_derived_assets.py, along with their associated config, ddl, and development files into a separate derived_assets_to_redshift folder
2. Creates new python virtual environment files that only use packages specific to asset_data_to_redshift and build_derived_assets
3. Edits the README.md file to removed details specific to the s3_to_redshift and cmslite_user_data_to_redshift microservices

Please note that some files have been modified for testing purposes. You can view the modified files in the ticket

### Login to the development environment:

1. Run the following command to login to EC2
```
$ awsmfa prod ######
Generating new IAM STS Token ...
STS Session Token generated and saved in profile mfa successfully.
$ microservice_ssm
```
2. Change the directory to the development branch:
```
$ cd /home/microservice/branch/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift
```

### Review README.md
Review [README.md](https://github.com/bcgov/GDX-Analytics-microservice/blob/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift/README.md) and check to makes sure there are no details specific to s3_to_redshift.py and cmslite_user_data_to_redshift.py. 
Please note that there are possibly still references to these other microservices where relevant. But these references shouldn't be related to how the microservice works or how the configs are structured

### Review moved files

Part of this changes includes moving files from the old combined folder into a separate one specific for this microservice. In this section, we will be reviewing to make sure all of these files were moved successfully
1. Review [derived_assets_to_redshift](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift) to make sure the microservice scripts have been moved successfully
- asset_data_to_redshift.py
- build_derived_assets.py
3. Run the following command to get a list of the scheduled microservice's configs:
```
$ crontab -l | grep build_derived_assets
```
4. Review the files in [config.d](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift/config.d) and make sure any config that appeared in the previous step is listed
5. Review the files in [ddl](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift/ddl) and make sure any table listed in one of the previous step's configs has an associated DDL
6. Review the files in [ddl](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift/ddl) to make sure the following DDLs exist
- derived.asset_downloads.sql
- derived.asset_downloads_derived.sql
- build_derived_assets.sql

### Review copied files

1. Review the files in [derived_assets_to_redshift](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift) to make sure the following file exist
- .flake8
- pylintrc
- README.md

### Review created files
1. Review the files in [derived_assets_to_redshift](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_derived_assets_to_redshift_split/derived_assets_to_redshift) to make sure the following file exist
- Pipfile
- Pipfile.lock

### Testing the microservice:
1. Run the following command to create the test tables:
```
$ microservice_rs < ddl/GDXDSD-7197_derived.asset_downloads.sql 
SET
CREATE TABLE
GRANT
$  microservice_rs < ddl/GDXDSD-7197_derived.asset_downloads_derived.sql 
SET
CREATE TABLE
GRANT
```
2. Run the following command to test the asset_data_to_redshift.py script and compare it to the expected output:
```
$ /home/microservice/.local/bin/pipenv run python3.8 asset_data_to_redshift.py config.d/GDXDSD-7197_intranet_assets.json
Report: asset_data_to_redshift.py

Config: config.d/GDXDSD-7197_intranet_assets.json

Microservice started at: 2024-11-21 09:49:18-0800 (PST), ended at: 2024-11-21 09:49:47-0800 (PST), elapsing: 0:00:29.293108.

Objects to process: 1
Objects successfully processed: 1
Objects that failed to process: 0
Objects output to 'processed/good': 1
Objects output to 'processed/bad': 0
Objects loaded to Redshift: 1
Empty Objects: 0

List of objects successfully fully ingested from S3, processed, loaded to S3 ('good'), and copied to Redshift:
1: client/doug_test/GDXDSD-7197/cmslite_gdx/intranet_apache.2024-11-02
```
3. Run the following query to check if the data was loaded correctly:
```
$ microservice_rs
SET
snowplow=> select count(*) from test.gdxdsd7197_asset_downloads;
 count 
-------
   380
(1 row)

snowplow=> \q
```
4. Run the following command to test the build_derived_assets.py script and compare it to the expected output:
```
$ /home/microservice/.local/bin/pipenv run python3.8 build_derived_assets.py config.d/GDXDSD-7197_intranet_assets.json

Report build_derived_assets.py:

Config: config.d/GDXDSD-7197_intranet_assets.json

Microservice started at: 2024-11-21 09:58:33-0800 (PST), ended at: 2024-11-21 09:58:53-0800 (PST), elapsing: 0:00:20.129298.

Objects to process: 1
Objects that failed to process: 0
Objects loaded to Redshift: 1

List of tables successfully parsed, and copied to the derived table in Redshift:
test.gdxdsd7197_asset_downloads

-----------------

```
5. Run the following query to check if the data was loaded correctly:
```
$ microservice_rs
SET
snowplow=> select count(*) from test.gdxdsd7197_asset_downloads_derived;
 count 
-------
   380
(1 row)

snowplow=> \q
```
6. Check the following folders in S3 to see if the intranet_apache.2024-11-02 file was successfully processed and appears in the following:
- batch: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/batch/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
- good: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/good/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
7. Delete the intranet_apache.2024-11-02 file from the following places in S3:
- batch: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/batch/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
- good: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/good/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
8. Run the following queries to drop the test tables:
```
$ microservice_rs
SET
snowplow=> drop table test.gdxdsd7197_asset_downloads;
DROP TABLE
snowplow=> drop table test.gdxdsd7197_asset_downloads_derived;
DROP TABLE
snowplow=> \q
```
